File: backend/src/main/java/com/spinwish/backend/models/requests/users/RegisterRequest.java
package com.spinwish.backend.models.requests.users;

import com.spinwish.backend.validators.ValidEmail;
import com.spinwish.backend.validators.ValidPassword;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.Pattern;
import jakarta.validation.constraints.Size;
import lombok.Data;

@Data
public class RegisterRequest {

    @NotBlank(message = "Email address is required")
    @ValidEmail
    private String emailAddress;

    @NotBlank(message = "Username is required")
    @Size(min = 3, max = 50, message = "Username must be between 3 and 50 characters")
    @Pattern(regexp = "^[a-zA-Z0-9_-]+$", message = "Username can only contain letters, numbers, underscores, and hyphens")
    private String username;

    @NotBlank(message = "Password is required")
    @ValidPassword
    private String password;

    @NotBlank(message = "Password confirmation is required")
    private String confirmPassword;

    @Pattern(regexp = "^\\+?[1-9]\\d{1,14}$", message = "Please enter a valid phone number")
    private String phoneNumber;

    @NotBlank(message = "Role name is required")
    private String roleName;
}

File: spinwishapp/lib/models/auth/register_request.dart
class RegisterRequest {
  final String emailAddress;
  final String username;
  final String password;
  final String confirmPassword;
  final String? phoneNumber;
  final String roleName;

  RegisterRequest({
    required this.emailAddress,
    required this.username,
    required this.password,
    required this.confirmPassword,
    this.phoneNumber,
    required this.roleName,
  });

  Map<String, dynamic> toJson() => {
        'emailAddress': emailAddress,
        'username': username,
        'password': password,
        'confirmPassword': confirmPassword,
        if (phoneNumber != null) 'phoneNumber': phoneNumber,
        'roleName': roleName,
      };

  factory RegisterRequest.fromJson(Map<String, dynamic> json) =>
      RegisterRequest(
        emailAddress: json['emailAddress'],
        username: json['username'],
        password: json['password'],
        confirmPassword: json['confirmPassword'],
        phoneNumber: json['phoneNumber'],
        roleName: json['roleName'],
      );
}
